gprofiler/python.py

```diff
@@ -188,40 +212,45 @@ def start(self):
             str(self._frequency),
             # Duration is irrelevant here, we want to run continuously.
         ]
-        process = start_process(cmd)
-        try:
-            poll_process(process, self.poll_timeout, self._stop_event)
-        except TimeoutError:
-            # Process is alive. So far, so good...
-            if not self.output_path.exists():
-                process.kill()
-                stdout = process.stdout.read().decode()
-                stderr = process.stderr.read().decode()
-                self._check_missing_headers(stdout)
-                raise CalledProcessError(process.returncode, process.args, stdout, stderr)
-            # happy flow
-            self.process = process
-        else:
-            # process terminated, must be an error:
-            self._check_output(process, self.output_path)
-            raise RuntimeError("_check_output did not raise!")
+        self.process = start_process(cmd)
+        # wait until the transient data file appears - because once returning from here, PyPerf may
+        # be polled via snapshot() and we need it to finish installing its signal handler.
+        if not wait_event(self.poll_timeout, self._stop_event, lambda: os.path.exists(self.output_path)):
+            raise TimeoutError("PyPerf didn't boot")
 
-    def _dump(self) -> Path:
-        assert self.process is not None, "profiling not started!"
-        self.process.send_signal(self.dump_signal)
+    def _glob_output(self) -> List[str]:
         # important to not grab the transient data file
-        while True:
-            output_files = glob.glob(f"{str(self.output_path)}.*")
-            if output_files:
-                break
+        return glob.glob(f"{str(self.output_path)}.*")
 
-            if self._stop_event.wait(0.1):
-                raise StopEventSetException()
+    def _wait_for_output_file(self, timeout: float) -> Optional[Path]:
+        if not wait_event(timeout, self._stop_event, lambda: len(self._glob_output()) > 0):
+            return None
 
+        output_files = self._glob_output()
         # All the snapshot samples should be in one file
         assert len(output_files) == 1
         return Path(output_files[0])
 
+    def _dump(self) -> Path:
+        assert self.process is not None, "profiling not started!"
+        self.process.send_signal(self.dump_signal)
+
+        path = self._wait_for_output_file(self.dump_timeout)
```

Comment:
Along with the above, it's clearer if we explicitly catch a `TimeoutError` here.

Reply:
Aight